### PR TITLE
Discovery Plus improvements

### PR DESCRIPTION
**Pull request by Ashish0804**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [ ] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [ ] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [ ] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Closes https://github.com/yt-dlp/yt-dlp/issues/1742
Closes https://github.com/yt-dlp/yt-dlp/issues/1574

Wasn't a rewrite as i'd thought of 😅 
has_drm wasn't set bcoz i couldn't find a url with DRM

### CODE REVIEW

File: yt_dlp/extractor/dplay.py [hunk 1]
```diff
@@ -426,3 +436,63 @@ class AnimalPlanetIE(DiscoveryPlusIE):
...
     _API_URL = 'us1-prod-direct.animalplanet.com'
+
+
+class DiscoveryPlusShowBaseIE(InfoExtractor):
+    def _entries(self, show_name):
+        headers = {
+            'x-disco-client': self._X_CLIENT,
+            'x-disco-params': 'realm=' + self._REALM,
+            'referer': self._DOMAIN,
+        }
```
Comment (pukkandan) [inline yt_dlp/extractor/dplay.py:447]: perhaps this can be a single class var

File: yt_dlp/extractor/dplay.py [hunk 2]
```diff
@@ -426,3 +436,63 @@ class AnimalPlanetIE(DiscoveryPlusIE):
...
+        headers['Authorization'] = 'Bearer ' + self._download_json(
+            self._BASE_API + 'token', show_name, 'Downloading token',
+            query={
+                'realm': self._REALM,
+                'deviceId': uuid.uuid4().hex,
+            })['data']['attributes']['token']
```
Comment (pukkandan) [inline yt_dlp/extractor/dplay.py:453]: does token have to be re-extracted for each playlist? why not move it into `_real_initialize`?

File: yt_dlp/extractor/dplay.py [hunk 3]
```diff
@@ -426,3 +436,63 @@ class AnimalPlanetIE(DiscoveryPlusIE):
...
+        show_url = self._BASE_API + 'cms/routes/' + self._SHOW_STR + '/{}?include=default'.format(show_name)
```
Comment (pukkandan) [inline yt_dlp/extractor/dplay.py:454]: fstrings can simply this a ton. similar below

File: yt_dlp/extractor/dplay.py [hunk 4]
```diff
@@ -426,3 +436,63 @@ class AnimalPlanetIE(DiscoveryPlusIE):
...
+                episodes_json = season_json['data']
+                for episode in episodes_json:
+                    video_id = episode['attributes']['path']
+                    yield self.url_result(
+                        '%svideos/%s' % (self._DOMAIN, video_id),
+                        ie=self._VIDEO_IE.ie_key(), video_id=video_id)
+                page_num += 1
+
+    def _real_extract(self, url):
+        show_name = self._match_valid_url(url).group('show_name')
```
Comment (pukkandan) [inline yt_dlp/extractor/dplay.py:478]: might as well use `_match_id`

**Review by pukkandan — COMMENTED** [review]
Not really a review. Just my thoughts while skimming through code

Also, consider moving everything to dplay.py

File: yt_dlp/extractor/dplay.py (hunk 4 above)
Comment (Ashish0804) [reply yt_dlp/extractor/dplay.py:478]: if this was spanning multiple lines for some reason, then yeah sure i would have left it as it is
But it's 1 line and more importantly allows dev to extend this to parse the domain if the _INDEX is same for 2 diff regions.

File: yt_dlp/extractor/dplay.py (hunk 2 above)
Comment (Ashish0804) [reply yt_dlp/extractor/dplay.py]: Can you review the headers handling again please?

**Pull request merged**